File: java/leetcode/src/Task14.java
import org.junit.jupiter.api.Test;

public class Task14 {
    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0) {
            return "";
        }

        String str = strs[0];
        int idx;
        for (idx = 0; idx < str.length(); idx++) {
            boolean same = true;
            for (String s : strs) {
                if (s.length() < idx + 1 || s.charAt(idx) != str.charAt(idx)) {
                    same = false;
                    break;
                }
            }
            if (!same) {
                break;
            }
        }

        return str.substring(0, idx);
    }

    public String longestCommonPrefix2(String[] strs) {
        if (strs.length == 0) {
            return "";
        }
        if (strs.length == 1) {
            return strs[0];
        }

        String str = strs[0];
        int left = 0, right = str.length() - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            String prefix = str.substring(0, mid + 1);
            boolean same = true;
            for (String s : strs) {
                if (!s.startsWith(prefix)) {
                    same = false;
                    break;
                }
            }
            if (same) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }


        return str.substring(0,left);
    }


    @Test
    void test() {
        String[] strs = new String[]{"flower","flow","flight"};
        System.out.println(longestCommonPrefix2(strs));
    }
}

File: py/Task14.py
# 编写一个函数来查找字符串数组中的最长公共前缀。

# 如果不存在公共前缀，返回空字符串 ""。

# 示例 1:

# 输入: ["flower","flow","flight"]
# 输出: "fl"
# 示例 2:

# 输入: ["dog","racecar","car"]
# 输出: ""
# 解释: 输入不存在公共前缀。
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        res = ''
        for itms in zip(*strs):
            if list(itms) == [itms[0]] * len(itms):
                res += itms[0]
            else:
                break
        return res

    def longestCommonPrefix2(self, strs: List[str]) -> str:
        res = ''
        for itms in zip(*strs):
            if len(set(itms)) == 1:
                res += itms[0]
            else:
                break
        return res


obj = Solution()
strs = ["flower", "flow", "flight"]
obj.longestCommonPrefix(strs)
